Components/CategoryResolver/DefaultCategoryResolver.php

```diff
@@ -16,26 +16,48 @@ class DefaultCategoryResolver extends CategoryResolver
      */
     public function resolve(array $categories, $shopId, $stream)
     {
-        $remoteCategoriesIds = $this->manager->getConnection()->executeQuery('
-            SELECT id
-            FROM s_plugin_connect_categories
-            WHERE shop_id = ? AND category_key IN (?)',
-            [$shopId, array_keys($categories)],
-            [\PDO::PARAM_INT, \Doctrine\DBAL\Connection::PARAM_STR_ARRAY])->fetchAll(\PDO::FETCH_COLUMN);
-
-        return $this->manager->getConnection()->executeQuery('
-            SELECT local_category_id
-            FROM s_plugin_connect_categories_to_local_categories
-            WHERE remote_category_id IN (?) AND (stream = ? OR stream IS NULL) ',
-            [$remoteCategoriesIds, $stream],
-            [\Doctrine\DBAL\Connection::PARAM_INT_ARRAY, \PDO::PARAM_STR])->fetchAll(\PDO::FETCH_COLUMN);
+        $tree = $this->generateTree($categories);
+
+        return array_keys($this->findAssignedLocalCategories($tree, $shopId, $stream));
     }
 
     /**
-     * {@inheritdoc}
+     * @param array $children
+     * @param int $shopId
+     * @param string $stream
+     * @return array
      */
-    public function generateTree(array $categories, $idPrefix = '')
+    private function findAssignedLocalCategories(array $children, $shopId, $stream)
     {
-        return [];
+        $childCategories = [];
+        $mappedCategories = [];
+        foreach ($children as $category) {
+            $localCategories = $this->manager->getConnection()->executeQuery('
+              SELECT pclc.local_category_id
+              FROM s_plugin_connect_categories_to_local_categories AS pclc
+              INNER JOIN s_plugin_connect_categories AS pcc ON pcc.id = pclc.remote_category_id
+              WHERE pcc.category_key = ? AND pcc.shop_id = ? AND (pclc.stream = ? OR pclc.stream IS NULL) ',
+              [$category['categoryId'], $shopId, $stream]
+            )->fetchAll(\PDO::FETCH_COLUMN);
+
+            if (is_array($localCategories)) {
```

Comment:
I meant the entire if is redundant:

```php
$data = (array) $this->giveMeAnArrayOrNull();

foreach($data as $key => $value) {
   ...
}
```
in this case the $data will always be an array

* `giveMeAnArrayOrNull` returns an array => $data is untouched
* `giveMeAnArrayOrNull` returns null => $data is an empty array with a count of 0

Reply:
does this work even is $data is false?
i dont know if there are differences between the different hydration modes 
but Connection::fetchColumn() will return false if the query doesn't match anything
that was the intent behind the original if